src/Moq/Interception/InterceptionAspects.cs

```diff
@@ -178,53 +178,47 @@ public override InterceptionAction Handle(Invocation invocation, Mock mock)
 			}
 
 			var methodName = invocation.Method.Name;
+			EventInfo eventInfo = null;
+			Action<string, Delegate> modifyEvent = null;
 
 			// Special case for event accessors. The following, seemingly random character checks are guards against
 			// more expensive checks (for the common case where the invoked method is *not* an event accessor).
 			if (methodName.Length > 4)
 			{
 				if (methodName[0] == 'a' && methodName[3] == '_' && invocation.Method.LooksLikeEventAttach())
 				{
-					var eventInfo = GetEventFromName(invocation.Method.Name.Substring("add_".Length), mock);
-					if (eventInfo != null)
-					{
-						// TODO: We could compare `invocation.Method` and `eventInfo.GetAddMethod()` here.
-						// If they are equal, then `invocation.Method` is definitely an event `add` accessor.
-						// Not sure whether this would work with F# and COM; see commit 44070a9.
+					eventInfo = GetEventFromName(invocation.Method.Name.Substring("add_".Length), mock);
+					modifyEvent = mock.EventHandlers.Add;
+				}
+				else if (methodName[0] == 'r' && methodName.Length > 7 && methodName[6] == '_' && invocation.Method.LooksLikeEventDetach())
+				{
+					eventInfo = GetEventFromName(invocation.Method.Name.Substring("remove_".Length), mock);
+					modifyEvent = mock.EventHandlers.Remove;
+				}
+			}
 
-						if (mock.CallBase && !invocation.Method.IsAbstract)
-						{
-							invocation.ReturnBase();
-							return InterceptionAction.Stop;
-						}
-						else if (invocation.Arguments.Length > 0 && invocation.Arguments[0] is Delegate delegateInstance)
-						{
-							mock.EventHandlers.Add(eventInfo.Name, delegateInstance);
-							invocation.Return();
-							return InterceptionAction.Stop;
-						}
+			if (eventInfo != null)
+			{
+				// TODO: We could compare `invocation.Method` and `eventInfo.GetRemoveMethod()` (or 
+				// `eventInfo.GetAddMethod()` here.
+				// If they are equal, then `invocation.Method` is definitely an event `remove` accessor.
+				// Not sure whether this would work with F# and COM; see commit 44070a9.
+
+				if (mock.CallBase && !invocation.Method.IsAbstract)
+				{
+					if (!mock.HasEventSetup)
```

Comment:
If I am understanding correctly, this means that when `mock.CallBase == true`, the base class' event accessor will no longer be called once an event (un-) subscription has been set up using `SetupSet`. This also means that an event (un-) subscription will be recorded like any other method call. Is this correct?

If so, then there's a piece missing here: You'd also need some `VerifySet` (or better, `VerifyAdd`/`VerifyRemove`) method that targets event handler (un-) subscriptions. Otherwise&mdash;unless you've marked the `SetupSet` [sic]  setup as `.Verifiable()`&mdash;you might run into situations where `VerifyNoOtherCalls` will always fail because you have unverifiable invocations to event accessors that you cannot reach using any of the other `Verify` methods.

That being said, I would prefer it even more if we did *not* have to add any more delegate-based `Setup` or `Verify` methods (such as `SetupSet` is), if possible. (The whole process behind it&mdash;"I'll try to figure out what the expression tree might have looked like by putting the mock in a dry-run state, then invoking the delegate and observing some side effects it causes"&mdash;strikes me as pretty messy and very limited, it just opens more doors for further bugs.)

Reply:
Yeah I missed that point. I now understand more deeply your concern. However I still see no way to bypass the C# limitation that event subscription cannot be used in expression tree because it is recognized as assignment.
Anyway, if we do go with the current proposed API, I'm not sure how to solve the particular problem you pointed on because I'm not familiar with the whole process of base classes.

Reply:
You're right, delegate-based methods would likely be unavoidable due to the limitations of LINQ expression trees.

Side note: I could provide some guidance regarding the event-related code paths & the `ReturnBase` stuff (which admittedly is not very easy to understand), for now the main thing is that we understand that additional `Verify{Add,Remove}` methods would likely be needed.